upgrade/upgrade-scripts/common.sh

```diff
@@ -362,3 +362,145 @@ function verify_upgrade_not_in_progress() {
 	. "$UPDATE_DIR/upgrade.properties" &>/dev/null
 	[[ -z "$UPGRADE_TYPE" ]] || die "upgrade currently in-progress"
 }
+
+function mask_service() {
+	local svc="$1"
+	local container="$2"
+
+	#
+	# Note that masking should succeed even if service doesn't exist
+	#
+	if [[ -n "$container" ]]; then
+		chroot "/var/lib/machines/$container" systemctl mask "$svc" ||
+			die "failed to mask '$svc' in container '$container'"
+	else
+		systemctl mask "$svc" || die "failed to mask '$svc'"
+	fi
+}
+
+function is_svc_masked_or_disabled() {
+	local svc="$1"
+
+	state=$(systemctl is-enabled "$svc")
+	if [[ "$state" == masked || "$state" == disabled ]]; then
+		return 0
+	fi
+
+	return 1
+}
+
+#
+# This function has 2 tasks:
+#  1. Fix/update the state of some services to be in line with what is expected
+#     in this version of the appliance.
+#  2. If we are doing a not-in-plaace upgrade, then migrate the state of the
+#     services into the upgrade container.
+#
+# It can be called from 2 different contexts:
+#  1. When creating upgrade container. In this case the container must be
+#     passed as first argument.
+#  2. When executing the in-place upgrade. In this case the function takes no
+#     arguments.
+#
+function fix_and_migrate_services() {
+	local container="$1"
+
+	#
+	# This function can be called from the execute script, however in that
+	# case we only want it to run when called on the running system and not
+	# in the upgrade container.
+	#
+	if systemd-detect-virt --container --quiet; then
+		echo "fix_and_migrate_services: should not run inside container"
+		return
+	fi
+
+	#
+	# We only want to run this logic on the Delphix Appliance variants.
+	# We especially want to avoid running it on the internal-dcenter
+	# variant, which has a different set of requirements for which
+	# services should be enabled / disabled. For instance, the
+	# dcenter-internal variant relies on either systemd-timesyncd.service
+	# or ntp.service to be enabled.
+	#
+	if ! dpkg -l delphix-virtualization &>/dev/null; then
+		echo "fix_and_migrate_services: should only run on the" \
+			"Delphix Appliance"
+		return
+	fi
+
+	#
+	# In versions prior to 6.0.13.0, snmpd.service was always enabled.
+	# Disable (mask) it here if we detect that it should have been disabled.
+	#
+	if compare_versions "$(get_current_version)" lt "6.0.13.0"; then
+		if [[ "$(systemctl is-enabled snmpd)" == enabled ]] &&
+			! grep -q "Delphix" /etc/snmp/snmpd.conf; then
+			mask_service snmpd "$container"
+		fi
+	fi
+
+	#
+	# Some services should be always masked.
+	#
+	while read -r svc; do
+		mask_service "$svc" "$container"
+	done <<-EOF
+		nginx.service
+		postgresql.service
+		systemd-timesyncd.service
```

Comment:
It feels like this list could live in one place (e.g. it could be a list in a file that is packaged/installed with the relevant variant). That way, the list isn't duplicated, and you also don't need the special code that skips this function for some variants?